### PR TITLE
update cookie sync endpoint to be case insensitive

### PR DESCRIPTION
**Pull request by gargcreation1992**
Tested locally with `/cookie_sync` POST endpoint with following body - 
```
{
"bidders": ["AppnexUs"],
"coopsync": false,
"limit": 1
}
```
and got this sample response - 
```
{
  "status" : "no_cookie",
  "bidder_status" : [
    {
      "no_cookie" : true,
      "bidder" : "AppnexUs",
      "usersync" : {
        "url" : "https://ib.adnxs.com/getuid?http%3A%2F%2Flocalhost%3A8000%2Fsetuid%3Fbidder%3Dappnexus%26gdpr%3D%26gdpr_consent%3D%26gpp%3D%26gpp_sid%3D%26f%3Di%26uid%3D%24UID",
        "type" : "redirect"
      }
    }
  ]
}
```

partially resolves https://github.com/prebid/prebid-server/issues/2400

### CODE REVIEW

File: usersync/chooser.go [hunk 1]
```diff
@@ -136,7 +140,12 @@ func (c standardChooser) Choose(request Request, cookie *Cookie) Result {
...
 func (c standardChooser) evaluate(bidder string, syncersSeen map[string]struct{}, syncTypeFilter SyncTypeFilter, privacy Privacy, cookie *Cookie) (Syncer, BidderEvaluation) {
-	syncer, exists := c.bidderSyncerLookup[bidder]
+	coreBidder, exists := c.normalizedBidderNamesLookup(bidder)
+	if !exists {
+		return nil, BidderEvaluation{Status: StatusUnknownBidder, Bidder: bidder}
+	}
+
+	syncer, exists := c.bidderSyncerLookup[coreBidder.String()]
 	if !exists {
 		return nil, BidderEvaluation{Status: StatusUnknownBidder, Bidder: bidder}
```
Comment (onkarvhanumante) [inline usersync/chooser.go:150]: Line 145 and 150 returns same error.  How about changing Line 150 to `syncerKeyNotFound` status

**Review by onkarvhanumante — COMMENTED** [review]
one suggestion. otherwise looks good to me 👍🏼

File: usersync/chooser.go (hunk 1 above)
Comment (gargcreation1992) [reply usersync/chooser.go:150]: `syncerKeyNotFound ` seems irrelevant to me here since this check is not related with syncher keys. 
This is actually checking if there is a coreBidder exists for the provided bidder in the request to which we are returning `UnknownBidder` if there is no coreBidder found for the provided bidder and that sounds ok to me.

File: endpoints/cookie_sync_test.go [hunk 2]
```diff
@@ -87,7 +87,7 @@ func TestNewCookieSyncEndpoint(t *testing.T) {
 	assert.IsType(t, &cookieSyncEndpoint{}, endpoint)
 
 	assert.Equal(t, expected.config, result.config)
-	assert.Equal(t, expected.chooser, result.chooser)
+	assert.ObjectsAreEqualValues(expected.chooser, result.chooser)
```
Comment (AlexBVolcy) [inline endpoints/cookie_sync_test.go:90]: For my own understanding, why was this changed?

Comment (gargcreation1992) [reply endpoints/cookie_sync_test.go:90]: When we mocked `normalizedBidderNamesLookup` field in standardChooser - apparently, go changes the address of the struct by copying the values to a new struct. Hence, the address are different however the values are same and that is what the assert statement is doing.

**Review by onkarvhanumante — DISMISSED** [review]

**Comment by gargcreation1992** [comment]
@bsardo Requesting you to please comment on the reason for the `blocked` label on this PR.

**Comment by bsardo** [comment]
This one is marked as blocked because it was determined that it conflicts with #3107 and the team came to the conclusion that we should try to merge that in first. It would be nice to get this in 2.0 but it isn't essential.

**Comment by SyntaxNode** [comment]
Unblocking. This is higher priority than the other PR.

**Review by SyntaxNode — COMMENTED** [review]
You also need to add case insensitivity for type filters and account for privacy checks in the usersyncPrivacy methods. Please add tests for those areas too.

**Comment by SyntaxNode** [comment]
Please also add a test for the chooser StatusDuplicate case, to make sure a duplicated bidder name triggers this code where the bidder name in the evaluated list matches what was provided in the request.

File: usersync/chooser.go [hunk 3]
```diff
@@ -136,7 +140,12 @@ func (c standardChooser) Choose(request Request, cookie *Cookie) Result {
 }
 
 func (c standardChooser) evaluate(bidder string, syncersSeen map[string]struct{}, syncTypeFilter SyncTypeFilter, privacy Privacy, cookie *Cookie) (Syncer, BidderEvaluation) {
-	syncer, exists := c.bidderSyncerLookup[bidder]
+	coreBidder, exists := c.normalizedBidderNamesLookup(bidder)
```
[the remark this answers is not in the document]
Comment (AlexBVolcy) [reply usersync/chooser.go:143]: There's an issue here where the original-cased bidder name is passed to `GDPRAllowsBidderSync` further below, which checks the original-cased bidder name against a list of all lowercased bidder names. So it means this function always returns false.

Should we just pass the `coreBidder` to these functions instead? To preserve the case insensitivity.

**Comment by AlexBVolcy** [comment]
I think there's a problem in `evaluate()`, I left a comment describing the issue.

File: usersync/chooser.go [hunk 4]
```diff
@@ -15,9 +17,10 @@ func NewChooser(bidderSyncerLookup map[string]Syncer) Chooser {
 	}
 
 	return standardChooser{
-		bidderSyncerLookup: bidderSyncerLookup,
-		biddersAvailable:   bidders,
-		bidderChooser:      standardBidderChooser{shuffler: randomShuffler{}},
+		bidderSyncerLookup:          bidderSyncerLookup,
+		biddersAvailable:            bidders,
+		bidderChooser:               standardBidderChooser{shuffler: randomShuffler{}},
+		normalizedBidderNamesLookup: openrtb_ext.NormalizeBidderName,
```
[the remark this answers is not in the document]
Comment (gargcreation1992) [reply usersync/chooser.go]: [c7685f9](https://github.com/prebid/prebid-server/pull/3103/commits/c7685f9c248c8346d28d4f44ba8a5f61b633abcf)

File: usersync/chooser.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gargcreation1992) [reply usersync/chooser.go]: [c7685f9](https://github.com/prebid/prebid-server/pull/3103/commits/c7685f9c248c8346d28d4f44ba8a5f61b633abcf)

File: usersync/chooser_test.go [hunk 5]
```diff
@@ -197,12 +211,30 @@ func TestChooserChoose(t *testing.T) {
...
+			bidderNamesLookup: func(name string) (openrtb_ext.BidderName, bool) {
+				return openrtb_ext.BidderName(name), false
+			},
+			expected: Result{
+				Status:           StatusOK,
+				BiddersEvaluated: []BidderEvaluation{{Bidder: "a", Status: StatusUnknownBidder}},
+				SyncersChosen:    []SyncerChoice{},
+			},
+		},
 	}
```
[the remark this answers is not in the document]
Comment (gargcreation1992) [reply usersync/chooser_test.go:268]: [526ce55](https://github.com/prebid/prebid-server/pull/3103/commits/526ce5507f25b1b81e086abcdf8c54147bb5f4e0)

File: usersync/chooser_test.go [hunk 6]
```diff
@@ -236,105 +269,130 @@ func TestChooserEvaluate(t *testing.T) {
...
+			givenSyncersSeen: map[string]struct{}{},
+			givenPrivacy:     fakePrivacy{gdprAllowsHostCookie: true, gdprAllowsBidderSync: true, ccpaAllowsBidderSync: true, activityAllowUserSync: true},
+			givenCookie:      cookieNeedsSync,
+			normalizedBidderNamesLookup: func(name string) (openrtb_ext.BidderName, bool) {
+				return openrtb_ext.BidderName(name), false
+			},
 			expectedSyncer:     nil,
-			expectedEvaluation: BidderEvaluation{Bidder: "a", SyncerKey: "keyA", Status: StatusBlockedByPrivacy},
+			expectedEvaluation: BidderEvaluation{Bidder: "a", Status: StatusUnknownBidder},
 		},
```
[the remark this answers is not in the document]
Comment (gargcreation1992) [reply usersync/chooser_test.go:478]: [526ce55](https://github.com/prebid/prebid-server/pull/3103/commits/526ce5507f25b1b81e086abcdf8c54147bb5f4e0)

**Comment by gargcreation1992** [comment]
> You also need to add case insensitivity for type filters and account for privacy checks in the usersyncPrivacy methods. Please add tests for those areas too.
For sync type filter - [88209d7](https://github.com/prebid/prebid-server/pull/3103/commits/88209d7ca5968facbcd42542217fa470d83ab3c0)
For UserSyncPrivacy - [176b887](https://github.com/prebid/prebid-server/pull/3103/commits/176b887993b70890769617a981d06369d961136a)

File: usersync/chooser.go [hunk 7]
```diff
@@ -147,7 +159,7 @@ func (c standardChooser) evaluate(bidder string, syncersSeen map[string]struct{}
 	}
 	syncersSeen[syncer.Key()] = struct{}{}
 
-	if !syncer.SupportsType(syncTypeFilter.ForBidder(bidder)) {
+	if !syncer.SupportsType(syncTypeFilter.ForBidder(strings.ToLower(bidder))) {
```
Comment (AlexBVolcy) [inline usersync/chooser.go:162]: Can't this line also use `bidderNormalized.String()`? And we don't have include the `strings` package into this file?

Comment (gargcreation1992) [reply usersync/chooser.go:162]: We cannot use normalised bidder name here and the reason is there is biddersLookup map that is created in `NewSpecificBidderFilter` function with [key of the map being lower cased. ](https://github.com/prebid/prebid-server/blob/88209d7ca5968facbcd42542217fa470d83ab3c0/usersync/bidderfilter.go#L47-L48). This map is used when we call `syncTypeFilter.ForBidder` function. Hope this clarifies.

Comment (AlexBVolcy) [reply usersync/chooser.go:162]: I see, but when I subbed in `bidderNomarlized.String()` for `strings.ToLower(bidder)`, all of the tests still passed. This is more of a nitpick if anything, so I'm happy to approve if you want to stick with the current code.

Comment (gargcreation1992) [reply usersync/chooser.go:162]: @AlexBVolcy It is because test cases are setup that way. If you setup a test case where you use `suntContent` as bidder name, you will be able to fail your test. In fact I have updated one of the test to reflect this - [c124abb](https://github.com/prebid/prebid-server/pull/3103/commits/c124abb52bfa929451685c5ee5a32569bba32f3f)

**Review by onkarvhanumante — APPROVED** [review]
